MINOR: ConfigKey variable renamed

kamalcph opened the pull request:
- Java doc updated
- Used final modifier where-ever it's applicable
- Unnecessary type-cast removed
- cleanup

kamalcph committed:
MINOR: Variable name corrected

- Java doc updated
- Used final modifier where-ever it's applicable
- Unnecessary type-cast removed
- cleanup

cmccabe commented:
LGTM (non-binding)

ijuma commented on clients/src/main/java/org/apache/kafka/common/utils/Utils.java:
```diff
@@ -334,14 +334,15 @@ public static void sleep(long ms) {
      */
     public static <T> T newInstance(Class<T> c) {
+        if (c == null)
+            throw new NullPointerException();
```
Why are we changing the exception type here? That could affect calling code. Also, if there's a good reason to do this, we'd want to include a message in the exception.

kamalcph replied:
I thought it's better to validate the argument before trying to create an instance. And, the code seems unfit in the `try..catch` block.

kamalcph replied:
This won't affect the calling code as both NPE and KafkaException extends RuntimeException. And, the message "Requested class was null" was dumb. 

I did this change, as other methods in this class followed the same pattern.

ijuma replied:
It makes a difference for code that catches `KafkaException` instead of `RuntimeException`. We'd have to check every usage to make sure it was OK.

kamalcph replied:
Makes sense. I'll revert this change.

ijuma commented on clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java:
```diff
@@ -158,8 +157,7 @@ public Password getPassword(String key) {
     /**
-     * Put all keys that do not start with {@code prefix} and their parsed values in the result map and then
-     * put all the remaining keys with the prefix stripped and their parsed values in the result map.
+     * Put all keys that start with {@code prefix} with the prefix stripped and their parsed values in the result map.
```
Why was this changed?

kamalcph replied:
The method doesn't process the keys that don't start with the prefix. So, I have updated the java doc.

ijuma replied:
Yes, it does. The copy constructor is used below.

```java
Map<String, Object> result = new RecordingMap<>(values(), prefix, true);
```

kamalcph replied:
Oh, I missed it.

ijuma commented on clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java:
```diff
@@ -66,52 +66,51 @@ public AbstractConfig(ConfigDef definition, Map<?, ?> originals) {
     }
 
-    protected Object get(String key) {
+    protected <T> T get(String key) {
```
I am not sure about this. It's a really dangerous method as it implicitly casts in an unsafe manner.

kamalcph replied:
I agree. Earlier, implicit casts were still done for primitives, list and class ?

ijuma replied:
We were doing explicit casts in those cases, right? I agree that it would be better to check before casting if that's what you are suggesting.

kamalcph replied:
Yes, I will do that change in another PR. I'll revert this one too.

ijuma reviewed (commented):
Thanks for the PR. It improves the code in a few areas. Left a few comments. It would also be good to update the PR title to reflect what's in the PR.

kamalcph committed:
Removed the implicit type-cast usages

- valuesWithPrefixOverride : javadoc restored to old state
- Utils#newInstance : actual behaviour restored

kamalcph commented:
@ijuma 

Thanks for the detailed review. I've updated this PR. Pl take a look.

ijuma reviewed (approved):
Thanks, LGTM.
